course6/lec6_script_en.md

```diff
@@ -0,0 +1,347 @@
+
+After that, our repeatedly additions are only the quick addition of new elements in `back`. Then, we remove elements from `front`. We check the invariant after the operation. We find that the queue is not empty, but `front` is empty, so we do retate `back` and move elements to `front` again. After that, we can normally take elements from `front`. 
+
+You can see that one rotation supports multiple removal operations, therefore the overall cost is much less than rebuilding the list every time.
+
+```moonbit no-check
+struct Queue[T] {
+  front: Stack[T]
+  back: Stack[T]
+}
+fn Queue::empty[T]() -> Queue[T] { {front: Empty, back: Empty} }
+
+// Store element at the end of the queue
+fn push[T](self: Queue[T], value: T) -> Queue[T] { 
+  normalize({ ..self, back: self.back.push(value)}) // By defining the first argument as self, we can use xxx.f()
+}
+
+// Remove the first element
+fn pop[T](self: Queue[T]) -> (Option[T], Queue[T]) { 
+  match self.front {
+    Empty => (None, self)
+    NonEmpty(top, rest) => (Some(top), normalize({ ..self, front: rest}))
+  }
+}
+
+// If front is empty, reverse back to front
+fn normalize[T](self: Queue[T]) -> Queue[T] { 
+  match self.front {
+    Empty => { front: self.back.reverse(), back: Empty }
+    _ => self
+  }
+}
+
+// Helper function: reverse the stack
+fn reverse[T](self: Stack[T]) -> Stack[T] { 
+  let mut result: Stack[T] = Empty
+  let mut current: Stack[T] = self
+  while let (Some(top), rest) = pop(current) {
+    result = result.push(top)
+    current = rest
+  }
+  // Implementation omitted
+}
+```
+
+Here is the code for the queue. You can see that we extensively apply generics, so our queue can contain any type, including queues containing other elements. The functions here are the specific implementations of the algorithm we just explained. In function `push`, you we called the stack's `push` function through `back.push`. We will explain this specifically in the next lesson.
+
+# Higher-Order Functions
+
+This section continues to focus on how to use the features provided by MoonBit to reduce repetitive code and enhance code reusability. So, let’s start with an example.
+
+```moonbit
+fn sum(list: List[Int]) -> Int {
+  match list {
+    Nil => 0
+    Cons(hd, tl) => hd + sum(tl)
+  }
+}
+```
+
+Consider some operations on lists. For instance, to sum an integer list, we use structural recursion with the following code: if empty, the sum is 0; if not empty, the sum is the current value plus the sum of the remaining list elements.
```

Comment:
```suggestion
Consider some operations on lists. For instance, to sum an integer list, we use structural recursion with the following code: if empty, the sum is 0; otherwise, the sum is the current value plus the sum of the remaining list elements.
```